### PR TITLE
Introduce MongoDB sharding rules to Documents and Document-wide collections

### PR DESCRIPTION
**Pull request by sejongk**
**What this PR does / why we need it**:
This PR introduces MongoDB sharding rules to `documents` and document-wide collections (e.g. `Changes`, `SyncedSync`, `Snapshots`).

**Which issue(s) this PR fixes**:

Addresses #673 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation**:

```docs

```

**Checklist**:
- [x] Added relevant tests or not required
- [x] Didn't break anything

### CODE REVIEW

**Review by hackerwins — COMMENTED** [review]
Thanks for your contribution.

I looked at the code up to the memory DB. and I have a question.

`MemoryDB` runs standalone, and it is not sharded based on shard keys. If we receive refKey due to change in interface, we only need to utilize the id. So there's no necessity to use the key in the implementation. What are your thoughts on this?

**Comment by sejongk** [comment]
> Thanks for your contribution.
> 
> I looked at the code up to the memory DB. and I have a question.
> 
> `MemoryDB` runs standalone, and it is not sharded based on shard keys. If we receive refKey due to change in interface, we only need to utilize the id. So there's no necessity to use the key in the implementation. What are your thoughts on this?

I think that's also a good idea. 
Previously, I heard that the `MemoryDB` is utilized for the testing purpose, so I tried to diminish the gap between the implementation of `MongoDB` and `MemoryDB` for the testing consistency.

File: server/backend/database/memory/indexes.go [hunk 1]
```diff
@@ -136,6 +147,16 @@ var schema = &memdb.DBSchema{
...
+				"project_id_id_key": {
+					Name: "project_id_id_key",
+					Indexer: &memdb.CompoundIndex{
+						Indexes: []memdb.Indexer{
+							&memdb.StringFieldIndex{Field: "ProjectID"},
+							&memdb.StringFieldIndex{Field: "ID"},
+							&memdb.StringFieldIndex{Field: "Key"},
+						},
+					},
+				},
```
[the remark this answers is not in the document]
Comment (sejongk) [reply server/backend/database/memory/indexes.go:159]: This index is used in https://github.com/yorkie-team/yorkie/pull/734/files#diff-18c3762643b65bd2b7add930409a5ae4d3b03add1e3340de78791bb8e6b9602bR843

[the remark this answers is not in the document]
Comment (hackerwins) [reply server/backend/database/memory/indexes.go:159]: Thank you for your answer. Are there any problems if we use `project_id_key_id` there?

File: test/helper/helper.go [hunk 2]
```diff
@@ -316,3 +323,163 @@ func NewRangeSlice(start, end int) []int {
...
+
+// CleanUpAllCollections removes all data in every collection.
+func CleanUpAllCollections(databaseName string) error {
+	cli, err := setupRawMongoClient(databaseName)
+	if err != nil {
+		return err
+	}
+
+	collections := []string{mongo.ColProjects, mongo.ColUsers, mongo.ColClients,
+		mongo.ColDocuments, mongo.ColChanges, mongo.ColSnapshots, mongo.ColSyncedSeqs}
```
Comment (hackerwins) [inline test/helper/helper.go:370]: When we add a new collection, it seems easy to forget to put it here.
How about `mongo` package exposing this array?

File: server/backend/database/memory/database.go [hunk 3]
```diff
@@ -1050,12 +1086,12 @@ func (d *DB) FindClosestSnapshotInfo(
...
 func (d *DB) FindMinSyncedSeqInfo(
 	_ context.Context,
-	docID types.ID,
+	docRefKey types.DocRefKey,
 ) (*database.SyncedSeqInfo, error) {
 	txn := d.db.Txn(false)
 	defer txn.Abort()
 
-	it, err := txn.Get(tblSyncedSeqs, "id")
+	it, err := txn.Get(tblSyncedSeqs, "doc_id_doc_key_client_id")
```
[the remark this answers is not in the document]
Comment (sejongk) [reply server/backend/database/memory/database.go:1094]: It seems this method is tested with MongoDB via the RPC testing. Adding test codes for this method to `Database` tests seems necessary.

**Review by hackerwins — APPROVED** [review]
LGTM. 👍

**Pull request merged**